### ISSUE TITLE
@angular@8.2.0-next.0: ERROR in ENOTDIR: not a directory, scandir *****/index.ts

### ISSUE THREAD

**Issue by julianobrasil** [angular/angular#31424]
# 🐞 bug report

### Affected Package
Angular 8.2.0-next.0

I wasn't able to compile an Ivy project using @angular@8.2.0-next.0 with Ivy enabled. It compiles using @angular 8.1.0

### Is this a regression?

Yes, the previous version in which this bug was not present was: Angular 8.1.0-next.3
### Description

I'm trying to compile an existing project (@nrwl recommended structure) using Ivy

## 🔥 Exception or Error
<pre><code>
ERROR in ENOTDIR: not a directory, scandir 'E:/Users/Juliano/Documents/Projetos/sa/sa-ui/saWorkspace/libs/issue-tracker/data-access/model/src/index.ts'

----------------------------------------------

10 silly lifecycle sa@0.0.0~build-pre: Args: [
10 silly lifecycle   '/d /s /c',
10 silly lifecycle   'ng build --configuration=preproduction --base-href /tutoria/'
10 silly lifecycle ]
11 silly lifecycle sa@0.0.0~build-pre: Returned: code: 1  signal: null
12 info lifecycle sa@0.0.0~build-pre: Failed to exec build-pre script
13 verbose stack Error: sa@0.0.0 build-pre: `ng build --configuration=preproduction --base-href /tutoria/`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Users\Juliano\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:200:13)
13 verbose stack     at ChildProcess.<anonymous> (C:\Users\Juliano\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:200:13)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
14 verbose pkgid sa@0.0.0
15 verbose cwd E:\Users\Juliano\Documents\Projetos\sa\sa-ui\saWorkspace
16 verbose Windows_NT 10.0.17134
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Juliano\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "run" "build-pre"
18 verbose node v12.3.1
19 verbose npm  v6.9.2
20 error code ELIFECYCLE
21 error errno 1
22 error sa@0.0.0 build-pre: `ng build --configuration=preproduction --base-href /tutoria/`
22 error Exit status 1
23 error Failed at the sa@0.0.0 build-pre script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]

</code></pre>

**Comment by michael42** [comment]
This might be the same issue that I get when setting `enableIvy: true` in my project, but I find the description here pretty vague.

`tsconfig.json`:
```
{
  "compilerOptions": {
     // ...
    "paths": {"config": ["src/config.dev.ts"]}
  },
  "angularCompilerOptions": { "enableIvy": true }
}
```

When building, this with Angular 8.1.0, I'm getting an`ERROR in ENOTDIR: not a directory, scandir '{...}/client/src/config.dev.ts'`, which makes me think there's an assumption that the path mappings always map to a directory. Setting an exception breakpoint allowed me to get the stack trace:
```
Error
    at handleErrorFromBinding (fs.js:126:5)
    at Object.readdirSync (fs.js:795:3)
    at NodeJSFileSystem.readdir (node_modules/@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.js:37:74)
    at EntryPointFinder.walkDirectoryForEntryPoints (node_modules/@angular/compiler-cli/ngcc/src/packages/entry_point_finder.js:85:18)
    at node_modules/@angular/compiler-cli/ngcc/src/packages/entry_point_finder.js:36:123
    at Array.reduce (<anonymous>)
    at EntryPointFinder.findEntryPoints (node_modules/@angular/compiler-cli/ngcc/src/packages/entry_point_finder.js:36:49)
    at Object.mainNgcc [as process] (node_modules/@angular/compiler-cli/ngcc/src/main.js:68:25)
    at NgccProcessor.processModule (node_modules/@ngtools/webpack/src/ngcc_processor.js:54:16)
    at node_modules/@ngtools/webpack/src/compiler_host.js:317:36
```

Maybe the `EntryPointFinder` should check if the path mappings do in fact map to a directory or not.

**Comment by gkalpak** [comment]
I think you are right, @michael42. The changes in #30525 affect how pathMappings are used, but doesn't solve the problem in its current state (see https://github.com/angular/angular/pull/30525/files#r300829346).

**Comment by spock123** [comment]
Getting same issue, for me it's the environment files that give IVY a headache..

**Comment by petebacondarwin** [comment]
This should be fixed as part of https://github.com/angular/angular/pull/30525

**Comment by julianobrasil** [comment]
@petebacondarwin, in another project of mine, I use `linqts`. It gives me a message slightly different: 

<pre><code>
ERROR in ENOENT: no such file or directory, lstat 'E:/Users/Juliano/Documents/Projetos/SA-R/sar-frontend/node_modules/linqts/index.ts'
</code></pre>

In fact, there's no `index.ts` on the folder mentioned on the error message. Is it a problem that I should post on `linqts` repo?

**Comment by petebacondarwin** [comment]
Yes this is a bug in the `package.json` for linqts. See https://unpkg.com/linqts@1.12.5/package.json

```
{
  "name": "linqts",
  "version": "1.12.5",
  "description": "LinQ + TypeScript",
  "es2015": "index.ts",
  "source": "index.ts",
  ...
```

The `es2015` property should not point to the non-existing `index.ts` file.

You can work around this by placing the following `ngcc.config.js` file at the root of your project:

```
module.exports = {
  packages: {
    'linqts': {
      entryPoints: {
        override: {
          es2015: undefined
        }
      }
    }
  }
};
```

This should tell ngcc to ignore this property in the `package.json`.

**Comment by julianobrasil** [comment]
Great! I'll open an issue on that repo and continue with ivy testing.

**Comment by julianobrasil** [comment]
Well, the compiler didn't like the undefined value: 

<pre><code>
ERROR in The "path" argument must be of type string. Received type object
</code></pre>

Is there a way to force it to accept it?

**Comment by petebacondarwin** [comment]
Interesting that it failed with the `undefined`... glad you got a workaround working.

**Comment by petebacondarwin** [comment]
Actually I just realised that `linqts` does not contain any Angular, does it? In which case we could just tell ngcc to ignore it:

```
module.exports = {
  packages: {
    'linqts': {
      entryPoints: {
        '.': {
          ignore: true
        }
      }
    }
  }
};
```

That being said, it should already be ignored by ngcc as it has no `...metadata.json` file, so I am not sure why it is erroring in the first place.

I tried:

```
mkdir temp
cd temp
yarn init
yarn add linqts
yarn add @angular/compiler
yarn add @angular/compiler-cli
yarn add typescript
yarn ivy-ngcc -l debug
```

which runs fine:

```
Compiling @angular/compiler/testing : fesm2015 as esm2015
Compiling @angular/compiler/testing : fesm5 as esm5
Compiling @angular/compiler/testing : esm2015 as esm2015
Compiling @angular/compiler/testing : esm5 as esm5
Compiling @angular/compiler/testing : main as umd
```

**Comment by petebacondarwin** [comment]
Ahah! But running it targeted on the `linqts` does fail:

```
yarn ivy-ngcc -l debug -t linqts
```

leads to 

```
Error: ENOENT: no such file or directory, lstat '...temp/node_modules/linqts/index.ts'
    at Object.lstatSync (fs.js:840:3)
    at NodeJSFileSystem.lstat (...temp/node_modules/@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.js:38:72)
    at CachedFileSystem.lstat (...temp/node_modules/@angular/compiler-cli/src/ngtsc/file_system/src/cached_file_system.js:87:38)
    at CachedFileSystem.readFile (...temp/node_modules/@angular/compiler-cli/src/ngtsc/file_system/src/cached_file_system.js:35:30)
    at EsmDependencyHost.recursivelyFindDependencies (...temp/node_modules/@angular/compiler-cli/ngcc/src/dependencies/esm_dependency_host.js:59:40)
    at EsmDependencyHost.findDependencies (...temp/node_modules/@angular/compiler-cli/ngcc/src/dependencies/esm_dependency_host.js:41:18)
    at DependencyResolver.getEntryPointDependencies (...temp/node_modules/@angular/compiler-cli/ngcc/src/dependencies/dependency_resolver.js:66:25)
    at TargetedEntryPointFinder.processNextPath (...temp/node_modules/@angular/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.js:57:42)
    at TargetedEntryPointFinder.findEntryPoints (...temp/node_modules/@angular/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.js:46:22)
    at getTargetedEntryPoints (...temp/node_modules/@angular/compiler-cli/ngcc/src/main.js:141:37)
```

**Comment by julianobrasil** [comment]
Good to know! After following you fixing that one, I found a dependency (`ng2-file-upload`), that became the next issue during compilation. It's maintained by valor-software. I'm reporting it to you here because it's kind of old (it uses @angular@2.x.x). Honestly, it seems to to be abandoned (for this Ivy testing I've just replaced it by an updated version made available by another user of that library). My question is: should I report to you guys this kind of thing (very old angular-based dependencies) or save your time and just replace it in these projects (it's not healthy to keep abandoned projects as dependencies anyway).
